### PR TITLE
Fix bug in update_process_group DDP API

### PR DESCRIPTION
**Pull request by pritamdamania87**
Fix bug in `_update_process_group` DDP API where we didn't correctly reset `local_used_map_` and a few other variables. This resulted in errors like `Encountered gradient which is undefined, but still allreduced by...`

Added a unit test as well that reproduced the issue.

cc @mrshenli @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -9888,6 +9888,57 @@ def test_ddp_update_process_group_new_group(self):
...
+
+                def forward(self, inp, error):
+                    if error:
+                        return self.fc3(self.fc2(self.fc1(SimulateError.apply(inp))))
+                    else:
+                        return self.fc2(self.fc1(inp))
+
+
+            input = torch.rand(10, 10, requires_grad=True).cuda(self.rank)
+            ddp = torch.nn.parallel.DistributedDataParallel(
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:9922]: To check my understanding, this DDP instance is still over the default global PG, and after erroring, the `ddp._update_process_group()` is reseting state but still using the global PG?

**Review by awgu — APPROVED** [review]
Reviewing as PT-D oncall 🫡

Idea makes sense to me!

**Review by fegin — APPROVED** [review]

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/distributed_test.py:9922]: Yes, we didn't need to use a different PG to reproduce the error.
